java/src/jmri/jmrix/roco/z21/Z21SensorManager.java

```diff
@@ -231,19 +232,27 @@ synchronized public String getNextValidAddress(String curAddress, String prefix)
      * {@inheritDoc}
      */
     @Override
-    public Sensor getBySystemName(String sName){
-       Sensor s = super.getBySystemName(sName);
-       if(s == null && sName.contains(":")) {
-          // normalize the hex characters in the system name to upper case.
-          String curAddress = sName.substring(getSystemPrefix().length() +1);
-          try {
-             return super.getBySystemName(createSystemName(curAddress, getSystemPrefix()));
-          } catch (JmriException je) {
-             // format isn't correct, but s already equals null, so just return
-             // that.
-          }
-       }
-       return s;
+    public Sensor getBySystemName(String sName) {
+        Sensor s = super.getBySystemName(sName);
+        if (s == null && sName.contains(":")) {
+            // normalize the hex characters in the system name to upper case.
+            String address = sName.substring(getSystemNamePrefix().length()).toUpperCase(Locale.ENGLISH);
```

Comment:
For Z21 CAN sensors and reporters, it isn’t the prefix that we are concerned with the case on.  

We are concerned that the node address in the system name, which is a hexadecimal number, is always equal regardless of case in the string.  I.e. ZSD123:1 and ZSd123:1 are equivalent, but they are not required to be the same as zSD123:1.

Reply:
That is why the prefix is stripped from sName before uppercasing.

Reply:
Ok, but this still makes this method more complex because you have to do this twice, once with uppercase and once with lower case.

The reason I was able to do this once with createSystemName is because createSystemName parses the hex value and then uses a String.format() to spit it back out.  This means the hex values are the same case regardless of the input.

If you're going to change this to makeSystemName, the Z21 code should override makeSystemName so that it preserves that behavior.

Reply:
Per the JMRI documentation, makeSystemName must accept, without modification, the name input by the user, so if the user wants to use lowercase, uppercase, or mixed case in a hexadecimal number, it must be accepted and used without modification or, if we want to require that hexadecimal numbers always be all uppercase or always be all lowercase, that can be done also. We **do not** normalize system names (other than prepending the prefix). Normalization was removed in 4.17.2; it appears that instances of createSystemName that normalize were missed when that was done.

Reply:
@rhwood the problem is that removing the normalization broke the Z21 code and it wasn’t detected because there were insufficient tests.

Part of the system name for the Z21 CAN Sensors and reporters Is a hexadecimal number.  Those numbers MUST be considered equal regardless of case.  We purposely are only normalizing the hexadecimal digits.

Why is this an issue?  Because we have to account for the possibility that a user has entered into a panel a sensor or reporter with a given case AND that we automatically generate Sensors and Reporters from messages received.  We have to account for the fact that these two processes can occur in either order and we can’t ( ever ) have two reporters or sensors when there should only be one.

This is a case where normalization of a substring of the System name IS important.  This isn’t applicable to most systems we have because those systems use decimal values.

I don’t care what case the name is actually in, I just want the hexadecimal values to produce the same result when we are trying t match them with registered sensors and reporters.

I.e.
`abcd == AbCd == aBcD == ABcd == ...`

Reply:
Hmmm.. looking at this further, this needs more work.

I found a test case in master where the Z21SensorManager does the right thing, but the Z21ReporterManager doesn't.  I think your version of this code will also fail that test for Z21SensorManager.

I'm going to re-implement getBySystemName without using either createSystemName or makeSystemName.